### PR TITLE
Improve rcl timer test coverage.

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says.

CI up to `rcl`:
* Linux (w/ coverage) [Build Status](http://ci.ros2.org/job/ci_linux/11183/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6462/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9134/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11138/)

### CODE REVIEW

File: rcl/test/rcl/test_timer.cpp [hunk 1]
```diff
@@ -183,7 +183,7 @@ TEST_F(TestTimerFixture, test_two_timers_ready_before_timeout) {
   rcl_timer_t timer2 = rcl_get_zero_initialized_timer();
 
   ret = rcl_timer_init(
-    &timer, &clock, this->context_ptr, RCL_MS_TO_NS(50), nullptr, rcl_get_default_allocator());
+    &timer, &clock, this->context_ptr, RCL_MS_TO_NS(10), nullptr, rcl_get_default_allocator());
```
Comment (brawner) [inline rcl/test/rcl/test_timer.cpp:186]: If this is an important change, can you add a comment so it's clearer why this value was chosen?

Comment (hidmic) [reply rcl/test/rcl/test_timer.cpp:267]: Dropped a comment in https://github.com/ros2/rcl/pull/680/commits/b19b6f078ac8da5d7f3a0086e9d7dc4b31c17b0e.

**Comment by brawner** [comment]
It looks like Windows fails on rmw_connext for `test_two_timers_ready_before_timeout`

Any feedback on whether you can add more coverage by using faulty clocks?

> If possible
> rcl_timer_call by setting timer->impl->clock as a function that returns negative values
> 
> It seems like you can also cause a lot of the rcl_clock_get_now calls to fail with an rcl_clock_t that doesn't have type or a get_now. Is that feasible to do?
>

**Comment by brawner** [comment]
Based on the coverage report, the faulty clocks aren't getting the remaining lines of coverage, or am I missing something?

https://ci.ros2.org/job/ci_linux_coverage/195/cobertura/src_ros2_rcl_rcl_src_rcl/timer_c/

**Comment by hidmic** [comment]
> It looks like Windows fails on rmw_connext for `test_two_timers_ready_before_timeout`

Hmm, yeah, again. It's too flaky. I'll disable timer tests against `rmw_connext_cpp` until https://github.com/ros2/rcl/issues/687 is resolved.

File: rcl/test/rcl/test_timer.cpp [hunk 2]
```diff
@@ -274,6 +357,104 @@ TEST_F(TestTimerFixture, test_timer_not_ready) {
...
+  // Force multiple timer timeouts.
+  ret = rcl_wait(&wait_set, RCL_MS_TO_NS(500));
+  EXPECT_EQ(RCL_RET_TIMEOUT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+
+  bool is_ready = false;
+  ret = rcl_timer_is_ready(&timer, &is_ready);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  EXPECT_TRUE(is_ready);
+  rcl_reset_error();
```
Comment (Blast545) [inline rcl/test/rcl/test_timer.cpp:419]: I think this one is not needed, same below for the RCL_RET_OK cases

**Comment by hidmic** [comment]
> Based on the coverage report, the faulty clocks aren't getting the remaining lines of coverage, or am I missing something?

Indeed. I was missing a couple places.

File: rcl/test/rcl/test_timer.cpp (hunk 2 above)
Comment (hidmic) [reply rcl/test/rcl/test_timer.cpp:419]: I put them just in case the expectation fails.

**Comment by hidmic** [comment]
Alright, [95% for `timer.c`](https://ci.ros2.org/job/ci_linux/11183/cobertura/src_ros2_rcl_rcl_src_rcl/timer_c/).

**Comment by hidmic** [comment]
@brawner PTAL

File: rcl/test/rcl/test_timer.cpp [hunk 3]
```diff
@@ -599,8 +798,15 @@ TEST_F(TestPreInitTimer, test_timer_call) {
...
 
+  EXPECT_EQ(RCL_RET_OK, rcl_enable_ros_time_override(&this->clock)) << rcl_get_error_string().str;
+  EXPECT_EQ(RCL_RET_OK, rcl_set_ros_time_override(&this->clock, -1)) << rcl_get_error_string().str;
+  EXPECT_EQ(RCL_RET_ERROR, rcl_timer_call(&timer));
+  rcl_reset_error();
+  EXPECT_EQ(times_called, 4);
+
   EXPECT_EQ(RCL_RET_OK, rcl_timer_cancel(&timer)) << rcl_get_error_string().str;
   EXPECT_EQ(RCL_RET_TIMER_CANCELED, rcl_timer_call(&timer));
+  rcl_reset_error();
```
[the remark this answers is not in the document]
Comment (hidmic) [reply rcl/test/rcl/test_timer.cpp:809]: Hmm, do you mean we ensure we don't have false `RCL_RET_OK`s?

**Comment by hidmic** [comment]
> I can't see the annotated lines because it looks new coverage jobs have been launched since this one, however it looks like it has jumped from 88% to 95% or so, which looks pretty good to me.

Yeah, that's a bummer. But I did check and I think we're as good as we can be w/o mocks/fault injection.

File: rcl/test/rcl/test_timer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (hidmic) [reply rcl/test/rcl/test_timer.cpp:797]: Still not sure, so I'm rolling those back. See https://github.com/ros2/rcl/pull/680/commits/6aa785ecfbad493be355f805101b27d1a719f954.

[the remark this answers is not in the document]
Comment (brawner) [reply rcl/test/rcl/test_timer.cpp:797]: Sorry I missed this comment. I meant to ensure that rcl_error_is_set() returns false before resetting it. That way if the error is set, we find out about it before resetting it for the next test segment.

[the remark this answers is not in the document]
Comment (hidmic) [reply rcl/test/rcl/test_timer.cpp:797]: Oh, but we would've found out about it when the success expectation is broken. In any case, I rolled it back. We can discuss whether we should be doing those checks irrespective of return codes on a follow-up issue/PR.

**Pull request merged**